### DISCUSSION TITLE
Classname confusion with conditionally rendered components in `_app`

### DISCUSSION THREAD

**Discussion by ky28059** [vercel/next.js#34680] [Help]
### Verify canary release

- [X] I verified that the issue exists in Next.js canary release

### Provide environment information

```
    Operating System:
      Platform: win32
      Arch: x64
      Version: Windows 10 Pro
    Binaries:
      Node: 16.7.0
      npm: 7.20.3
      Yarn: N/A
      pnpm: N/A
    Relevant packages:
      next: 12.1.0
      react: 17.0.2
      react-dom: 17.0.2
```

### What browser are you using? (if relevant)

Chrome 98.0.4758.102

### Describe the Bug

Passing `<Component {...pageProps} />` as children to `<Layout>` causes the page's classnames to be mixed up with that of `<Sidebar />`, which is rendered alongside `props.children` in the layout.

![image](https://user-images.githubusercontent.com/60120929/154782163-756de13b-4b59-4f62-853e-43880d969ba3.png)
![image](https://user-images.githubusercontent.com/60120929/154782954-77147c27-78be-4427-a82d-82ae7fdc436a.png)

`Layout.tsx`:
```tsx
import {useContext, useEffect, ReactNode} from 'react';
// ...
import {useScreenType} from '../hooks/useScreenType';
// ...
import Sidebar from './layout/Sidebar';
import BottomNav from './layout/BottomNav';


export default function Layout(props: {children: ReactNode}) {
    // ...

    return (
        <>
            <div id="app" className={screenType === 'phone' ? 'vertical' : ''}>
                {screenType !== 'phone' && <Sidebar forceCollapsed={screenType === 'smallScreen'} />}
                <div id="content">
                    {props.children}
                </div>
                {screenType === 'phone' && <BottomNav />}
            </div>
            <SgyInitResults />
        </>
    );
}
```
`Sidebar.tsx`:
```tsx
type SidebarProps = {forceCollapsed?: boolean};
export default function Sidebar(props: SidebarProps) {
    // ...

    return (
        <div className={`sidebar ${!isOpen ? 'collapsed' : ''}`}>
            {/* The outer <div> only exists to occupy space to squish #content; the inner <div> has position: fixed; */}
            <div className="sidebar-content">
                {/* ... */}
            </div>
        </div>
    )
}
```
`_app.tsx`:
```tsx
import {AppProps} from 'next/app';


export default function App({ Component, pageProps }: AppProps) {
    // ...

    return (
        {/* ... */}
            <Layout>
                <Component {...pageProps} />
            </Layout>
        {/* ... */}
    );
}
```
`index.tsx`:
```tsx
import {useScreenType} from '../hooks/useScreenType';


export default function Home() {
    // ...

    // Screen type for responsive design
    const screenType = useScreenType();
    const displayFromScreenType = () => {
        if (screenType === 'phone') return 'one-col narrow';
        if (screenType === 'smallScreen') return 'one-col';
        return 'two-col';
    }
    // ...

    return (
        <div className={`home ${displayFromScreenType()}`}>
            {/* ... */}
        </div>
    );
}
```
When rendering, the page wrapper's `id="content"` is given to `Sidebar`'s enclosing div, and the classnames of the rendered page override those of `Sidebar`'s inner content div. This behavior persists after rerunning `npm run dev` and deleting the `.next` folder.

Strangely, copying and pasting the layout from `Layout` directly to `_app` fixes the issue.

![image](https://user-images.githubusercontent.com/60120929/154782847-5486c5ba-3480-4615-81f9-96342f7e594d.png)
![image](https://user-images.githubusercontent.com/60120929/154782863-3b4759da-3e64-48d8-93c5-384cc3ff19da.png)

Replacing the copied layout to again use `<Layout>` causes `Layout` to function correctly after `npm run dev`'s automatic refresh, but force refreshing the page gives the issue again.

### Expected Behavior

I expected the classnames declared in `<Sidebar />` to not be overridden by the classnames of the enclosing `<Layout>` and page.

### To Reproduce

https://github.com/GunnWATT/watt/tree/next-port/client

**Comment by ky28059** [comment]
Issue also exists on Next `12.0.10`, if that helps.

**Comment by ky28059** [comment]
![image](https://user-images.githubusercontent.com/60120929/154813792-a606d1d7-fa47-407e-985b-e4f0ff22762c.png)
![image](https://user-images.githubusercontent.com/60120929/154813803-b4047f4a-a322-449b-8586-f054b9d722bf.png)

Hmm, still happens when the layout is fully replicated in `_app`. It's something to do with the conditionals.

**Comment by ky28059** [comment]
![image](https://user-images.githubusercontent.com/60120929/154814252-0af44906-9930-4d1e-9b84-c8c8c83a3ef9.png)
![image](https://user-images.githubusercontent.com/60120929/154814279-114ed880-3d2a-4b2f-a564-641e046b9a70.png)
![image](https://user-images.githubusercontent.com/60120929/154814642-550263e6-e215-4a62-9bc4-c4f85e3e45f5.png)

Also happens when the screenType conditional is moved from `Layout` to `Sidebar` and `Sidebar` is always rendered.

**Comment by macrozone** [comment]
i guess your problem is that you a hook `useScreenType` to change the layout, which won't work with Server Side Rendering.  So your server does not know what device the client is and will allways render one version. 

That could lead to hydration problems like props beeing assigned to the wrong element.
